File: backend/src/usage/usage.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { UsageEvent } from '../common/entities/usage-event.entity';

@Injectable()
export class UsageService {
  constructor(
    @InjectRepository(UsageEvent)
    private readonly usageRepository: Repository<UsageEvent>
  ) {}

  async record(event: Partial<UsageEvent>): Promise<UsageEvent> {
    const usage = this.usageRepository.create(event);
    return this.usageRepository.save(usage);
  }

  async getDailyTotals(tenantId: string, date = new Date()) {
    const start = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), 0, 0, 0));
    const end = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate() + 1, 0, 0, 0));

    const result = await this.usageRepository
      .createQueryBuilder('usage')
      .select('SUM(usage.tokensIn + usage.tokensOut)', 'tokens')
      .addSelect('SUM(usage.costUsd)', 'costUsd')
      .where('usage.tenantId = :tenantId', { tenantId })
      .andWhere('usage.createdAt >= :start', { start })
      .andWhere('usage.createdAt < :end', { end })
      .getRawOne();

    return {
      tokens: Number(result?.tokens || 0),
      costUsd: Number(result?.costUsd || 0)
    };
  }

  async getSummaryByTenant(tenantId: string) {
    const totals = await this.getDailyTotals(tenantId);
    return {
      tenantId,
      tokens: totals.tokens,
      costUsd: totals.costUsd
    };
  }

  async getSummaryAll(date = new Date()) {
    const start = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), 0, 0, 0));
    const end = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate() + 1, 0, 0, 0));
    const result = await this.usageRepository
      .createQueryBuilder('usage')
      .select('usage.tenantId', 'tenantId')
      .addSelect('SUM(usage.tokensIn + usage.tokensOut)', 'tokens')
      .addSelect('SUM(usage.costUsd)', 'costUsd')
      .where('usage.createdAt >= :start', { start })
      .andWhere('usage.createdAt < :end', { end })
      .groupBy('usage.tenantId')
      .getRawMany();

    return result.map((row) => ({
      tenantId: row.tenantId,
      tokens: Number(row.tokens || 0),
      costUsd: Number(row.costUsd || 0)
    }));
  }

  async listEvents(tenantId: string, limit = 20) {
    return this.usageRepository.find({
      where: { tenantId },
      order: { createdAt: 'DESC' },
      take: limit
    });
  }

  async listEventsAll(limit = 20) {
    return this.usageRepository.find({
      order: { createdAt: 'DESC' },
      take: limit
    });
  }
}

File: backend-java/src/main/java/com/neria/manager/usage/UsageService.java
package com.neria.manager.usage;

import com.neria.manager.common.entities.UsageEvent;
import com.neria.manager.common.repos.UsageEventRepository;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class UsageService {
  private final UsageEventRepository usageRepository;

  public UsageService(UsageEventRepository usageRepository) {
    this.usageRepository = usageRepository;
  }

  public UsageEvent record(UsageEvent event) {
    if (event.getId() == null || event.getId().isBlank()) {
      event.setId(UUID.randomUUID().toString());
    }
    if (event.getCreatedAt() == null) {
      event.setCreatedAt(LocalDateTime.now());
    }
    if (event.getCostUsd() == null) {
      event.setCostUsd(BigDecimal.ZERO);
    }
    return usageRepository.save(event);
  }

  public Totals getDailyTotals(String tenantId, LocalDate date) {
    LocalDateTime start = date.atStartOfDay();
    LocalDateTime end = date.plusDays(1).atStartOfDay();
    List<UsageEvent> events =
        usageRepository.findTop200ByTenantIdOrderByCreatedAtDesc(tenantId);
    int tokens = 0;
    double cost = 0;
    for (UsageEvent event : events) {
      if (event.getCreatedAt() == null) {
        continue;
      }
      LocalDateTime createdAt = event.getCreatedAt();
      if (createdAt.isBefore(start) || !createdAt.isBefore(end)) {
        continue;
      }
      tokens += event.getTokensIn() + event.getTokensOut();
      cost += event.getCostUsd().doubleValue();
    }
    return new Totals(tokens, cost);
  }

  public Map<String, Object> getSummaryByTenant(String tenantId) {
    Totals totals = getDailyTotals(tenantId, LocalDate.now(ZoneOffset.UTC));
    return Map.of("tenantId", tenantId, "tokens", totals.tokens, "costUsd", totals.costUsd);
  }

  public List<Map<String, Object>> getSummaryAll() {
    List<UsageEvent> events = usageRepository.findTop200ByOrderByCreatedAtDesc();
    LocalDate today = LocalDate.now(ZoneOffset.UTC);
    LocalDateTime start = today.atStartOfDay();
    LocalDateTime end = today.plusDays(1).atStartOfDay();
    return events.stream()
        .filter(event -> event.getCreatedAt() != null)
        .filter(event -> !event.getCreatedAt().isBefore(start) && event.getCreatedAt().isBefore(end))
        .collect(java.util.stream.Collectors.groupingBy(UsageEvent::getTenantId))
        .entrySet()
        .stream()
        .map(
            entry -> {
              int tokens = 0;
              double cost = 0;
              for (UsageEvent event : entry.getValue()) {
                tokens += event.getTokensIn() + event.getTokensOut();
                cost += event.getCostUsd().doubleValue();
              }
              Map<String, Object> summary = new java.util.HashMap<>();
              summary.put("tenantId", entry.getKey());
              summary.put("tokens", tokens);
              summary.put("costUsd", cost);
              return summary;
            })
        .toList();
  }

  public List<UsageEvent> listEvents(String tenantId, int limit) {
    List<UsageEvent> events = usageRepository.findTop200ByTenantIdOrderByCreatedAtDesc(tenantId);
    return events.subList(0, Math.min(limit, events.size()));
  }

  public List<UsageEvent> listEventsAll(int limit) {
    List<UsageEvent> events = usageRepository.findTop200ByOrderByCreatedAtDesc();
    return events.subList(0, Math.min(limit, events.size()));
  }

  public record Totals(int tokens, double costUsd) {}
}
